### PR TITLE
[BEAM-999] Block earlier in BoundedReadEvaluatorFactoryTest

### PR DESCRIPTION
**Pull request by tgroh**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

This ensures that the reader doesn't claim the split point, which in
turn ensures the dynamic split request will not be refused by the
OffsetBasedSource. If the split is refused, ...ProducesDynamicSplits
flakes, as if the reader is faster than the split thread it can run past
the point at which the splitter thread will attempt to split the source,
which causes the reader to read all of the elements.

Spin within TestReader#advanceImpl if the reader is being dynamically
split, to ensure that the dynamic split fully completes before
continuing a call to advance.

### CODE REVIEW

**Comment by tgroh** [comment]
R: @dhalperi OR @kennknowles OR @lukecwik

**Comment by tgroh** [comment]
I ran the flaky test with changes 10k times with no failures.

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/BoundedReadEvaluatorFactoryTest.java [hunk 1]
```diff
@@ -434,9 +432,13 @@ public boolean startImpl() throws IOException {
 
     @Override
     public boolean advanceImpl() throws IOException {
-      if (index == sleepIndex) {
+      // Sleep before the sleep/split index is claimed so long as it will be claimed
+      if (index + 1 == sleepIndex && sleepIndex < getCurrentSource().elems.length) {
         try {
           dynamicallySplit.await();
+          while (initialSource.equals(getCurrentSource())) {
+            // Spin until the current source is updated
```
Comment (kennknowles) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/BoundedReadEvaluatorFactoryTest.java]: Note sure I've grokked everything, but I'd love an active way to receive this message. Perhaps override `splitAtFraction` and send the signal after the call to `super.splitAtFraction` completes?

Comment (tgroh) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/BoundedReadEvaluatorFactoryTest.java]: `OffsetBasedSource#splitAtFraction` is final, so we can't do it there. There also isn't any reader method called after the source is updated, so we can't add a hook there.

I've added a sleep of 1ms to make sure the CPU is yielded here.

**Comment by kennknowles** [comment]
LGTM

**Pull request closed without merging**